### DISCUSSION TITLE
How build  binaries using Visual Studio (2017+)?

### DISCUSSION THREAD

**Discussion by neolib** [rikyoz/bit7z#176] [Q&A] [answered]
I see that version 4 uses cmake files, but I am not familiar with cmake build system. Can anyone show me how to build binaries from source using Visual Studio?

**Comment by rikyoz** [comment, marked as answer]
Hi!

### Directly in Visual Studio

In Visual Studio 2017, you can directly open CMake projects by going to the `File` menu > `Open` > `CMake...`, and then opening the `CMakeList.txt` file of bit7z:
![image](https://github.com/rikyoz/bit7z/assets/1334291/bb56ea2e-428e-40fb-a280-c7eca1d70123)
![image](https://github.com/rikyoz/bit7z/assets/1334291/d9ab6128-51b9-4eb8-b81a-77be86c68355)

### Using the CMake's GUI

Alternatively, you can use the CMake GUI to generate a Visual Studio project for bit7z, which you can then use to build the library:

![image](https://github.com/rikyoz/bit7z/assets/1334291/25c7fa77-9e9a-47e8-8d5a-ab695faff243)
After selecting bit7z's source code path and the path where the generated Visual Studio project will be, you can click the "Configure" button.

A window will appear where you can select the project generator (in your case, `Visual Studio 15 2017`) as well as the platform (`Win32` for 32-bit, `x64` for 64-bit):
![image](https://github.com/rikyoz/bit7z/assets/1334291/405588e8-1ef4-4e8c-9d4e-b0a17c50c335)

CMake will then configure the project and will make available all the bit7z build options:
![image](https://github.com/rikyoz/bit7z/assets/1334291/3d5cb887-e5cc-47b8-ba2b-0bc5e37d7811)

After selecting the options you want to use for building bit7z, you can click the "Generate" button to make CMake generate the Visual Studio project and solution files in the specified path. You can manually open them in Visual Studio or click CMake's "Open Project" button.

![image](https://github.com/rikyoz/bit7z/assets/1334291/9527376f-bb18-4d26-aa2e-c38b114a6b6c)

### Using CMake's CLI

An alternative to using CMake's GUI is CMake's CLI.
In this case, you create the directory where you want to generate the Visual Studio project and, in a terminal inside of the directory, you run the command:
`cmake -G "Visual Studio 17 2022" -A "x64" <path-to-bit7z> <bit7z-options>`
where bit7z options are specified as `-DOPTION_NAME=<ON|OFF>` (e.g., `-DBIT7Z_AUTO_FORMAT=ON` enables the option `BIT7Z_AUTO_FORMAT`).
